compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -2626,8 +2627,10 @@ impl<'tcx> ProvisionalEvaluationCache<'tcx> {
     fn on_completion(&self, dfn: usize) {
         debug!(?dfn, "on_completion");
 
+        // If IndexMap gets the `drain_filter` method this could be done in one iteration
+        self.map.borrow_mut().retain(|_k, eval| eval.from_dfn < dfn);
         for (fresh_trait_pred, eval) in
-            self.map.borrow_mut().drain_filter(|_k, eval| eval.from_dfn >= dfn)
+            self.map.borrow().iter().filter(|(_k, eval)| eval.from_dfn >= dfn)
         {
             debug!(?fresh_trait_pred, ?eval, "on_completion");
         }
```

Comment:
This is now a different behavior as the map is cleared of elements matching the predicate.

Reply:
Yes, I messed up the order. Thanks for catching that!

Reply:
Should be fixed now!